File: src/main/java/com/cyhone/Cmd.java
package com.cyhone;

import org.apache.commons.cli.*;

import java.util.Arrays;
import java.util.List;

/**
 * @author cyhone
 * @date 2017/2/14
 */
public class Cmd {
    private Options options;
    private CommandLine cmd;
    private String[] args;

    public Cmd(String[] args) throws ParseException {
        options = new Options();
        init();
        parse(args);
    }

    // java [-options] class [args...]
    void init() {

        // help
        options.addOption("help", false, "help information");
        options.addOption("?", false, "help information");

        //version
        options.addOption("version", false, "print version and exit");

        //classpath
        Option cpOption = Option.builder("classpath")
                .required(false)
                .desc("classpath")
                .hasArg()
                .argName("path")
                .build();
        options.addOption(cpOption);

        Option cpOptionAbbr = Option.builder("cp")
                .required(false)
                .desc("classpath")
                .hasArg()
                .argName("path")
                .build();
        options.addOption(cpOptionAbbr);

        //Xjre
        Option xjreOption = Option.builder("Xjre")
                .required(false)
                .desc("path to jre")
                .hasArg()
                .argName("Xjre")
                .build();
        options.addOption(xjreOption);

    }

    public boolean isHelp() {
        return cmd.hasOption("help") || cmd.hasOption("?");
    }

    public boolean isVersion() {
        return cmd.hasOption("version");
    }

    public void printUsage() {
        HelpFormatter hf = new HelpFormatter();
        String formatstr = "java  [-options] class [args...]";
        hf.printHelp(formatstr, "", this.options, "");
    }

    public String getXjrePath() {
        if (cmd.hasOption("Xjre")) {
            return cmd.getOptionValue("Xjre");
        }
        return "";

    }

    public String getCpPath() {
        if (cmd.hasOption("classpath")) {
            return cmd.getOptionValue("classpath");
        }

        if (cmd.hasOption("cp")) {
            return cmd.getOptionValue("cp");
        }

        return "";
    }

    public String getClassName() {
        return args[0];
    }

    public void printVersion() {
        System.out.println("version 0.0.1");
    }

    public void parse(String[] args) throws ParseException {
        CommandLineParser parser = new DefaultParser();
        cmd = parser.parse(options, args);
        this.args = args.clone();

    }
}

File: src/cmd.cpp
#include "cmd.h"

namespace cyh {
    CmdParser::
    CmdParser() {
        init();
    }

    CmdParser::
    CmdParser(int argc, char **argv) : CmdParser() {
        parser(argc, argv);
    }

    void CmdParser::
    parser(int argc, char **argv) {
        cmdParser.parse_check(argc, argv);
        className = std::string(argv[1]);
    }

    bool CmdParser::
    isVersion() {
        return cmdParser.exist("version");
    }

    std::string CmdParser::
    xjreOption() {
        return cmdParser.get<std::string>("Xjre");
    }

    std::string CmdParser::cpOption() {
        auto result = cmdParser.get<std::string>("cp");
        return result.empty() ? cmdParser.get<std::string>("classpath") : result;
    }

    std::string CmdParser::
    getClassName() {
        return className;
    }

    void CmdParser::
    init() {
        //version
        cmdParser.add("version", '\0', "print version and exit");

        //classpath
        cmdParser.add<std::string>("classpath", '\0', "classpath", false, "");
        cmdParser.add<std::string>("cp", '\0', "classpath", false, "");

        //Xjre
        cmdParser.add<std::string>("Xjre", '\0', "path to jre", false, "");
    }

    void printVersion() {
        std::cout << "version 0.0.1" << std::endl;
    }
}

